## **1. 📊 Introduction to Pandas**

Pandas is a powerful library for data manipulation and analysis in Python. It builds on top of <font color="greed">NumPy</font> and provides easy-to-use data structures and functions for handling structured data. 🚀

### **Why Use Pandas?**
1. **Real-World Relevance**: Pandas helps analyze large datasets efficiently, such as customer reviews, financial data, or restaurant data like our **Zomato Bangalore dataset**.  
2. **Core Strengths**:  
   - Data cleaning, preparation, and exploration.
   - Supports handling **missing data**.  
   - Integrates seamlessly with libraries like <font color="magenta">Matplotlib</font> and <font color="magenta">NumPy</font>.  




**Installing Pandas**
To install Pandas, use the following command:

In [ ]:
!pip install pandas

💡 Tip: Pandas is often pre-installed in environments like Google Colab or Jupyter Notebooks.

### Loading the Dataset

Let's load our Zomato Bangalore dataset.


In [3]:
import requests
import zipfile
import os

url = "https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/100/375/original/zomato_dataset.zip"
zip_name = "zomato_dataset.zip"

# 1. Download
print("Downloading...")
response = requests.get(url, stream=True)
with open(zip_name, 'wb') as f:
    f.write(response.content)

# 2. Unzip
print("Unzipping...")
with zipfile.ZipFile(zip_name, 'r') as zip_ref:
    zip_ref.extractall(".")

# 3. Cleanup
os.remove(zip_name)
print("Done! Dataset is ready.")

Downloading...
Unzipping...
Done! Dataset is ready.


In [4]:
import pandas as pd

df = pd.read_csv('zomato_dataset.csv')

# Let's look at the first few rows to understand the structure
df.head()

,url,address,name,online_order,book_table,rate,votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city)
0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Jalsa,Yes,Yes,4.1/5,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,4.1/5,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari
2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,3.8/5,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari
3,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",Addhuri Udupi Bhojana,No,No,3.7/5,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari
4,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",Grand Village,No,No,3.8/5,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari


In [5]:
df.tail()

,url,address,name,online_order,book_table,rate,votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city)
23188,https://www.zomato.com/bangalore/izakaya-gastr...,"2nd Floor, Iona, Virginia Mall, Whitefield Mai...",Izakaya Gastro Pub,Yes,Yes,3.8 /5,128,+91 7625087121\n+91 8050587483,Whitefield,"Bar, Casual Dining","Beer, Chicken Guntur, Paneer Tikka, Fish, Nood...","North Indian, Continental, Mediterranean",'1200.0',"[('Rated 3.0', ""RATED\n Nice place to hangout...",[],Pubs and bars,Whitefield
23189,https://www.zomato.com/bangalore/m-bar-bengalu...,"Bengaluru Marriott Hotel, 75, 8th Road, EPIP A...",M Bar - Bengaluru Marriott Hotel Whitefield,No,No,3.9 /5,77,080 49435000,Whitefield,"Fine Dining, Bar",Rooftop Ambience,Finger Food,'2000.0',"[('Rated 4.0', 'RATED\n Went there post dinne...",[],Pubs and bars,Whitefield
23190,https://www.zomato.com/bangalore/keys-cafe-key...,"Keys Hotel, 6, 1st Phase Industrial Area, Near...",Keys Cafe - Keys Hotel,No,No,2.8 /5,161,080 39451000\n+91 8884038484,Whitefield,"Casual Dining, Bar","Salads, Coffee, Breakfast Buffet, Halwa, Chick...","Chinese, Continental, North Indian",'1200.0',"[('Rated 3.0', 'RATED\n Place is good not tha...",[],Pubs and bars,Whitefield
23191,https://www.zomato.com/bangalore/bhagini-2-whi...,"139/C1, Next To GR Tech Park, Pattandur Agraha...",Bhagini,No,No,2.5 /5,81,080 65951222,Whitefield,"Casual Dining, Bar","Biryani, Andhra Meal","Andhra, South Indian, Chinese, North Indian",'800.0',"[('Rated 4.0', 'RATED\n A fine place to chill...",[],Pubs and bars,Whitefield
23192,https://www.zomato.com/bangalore/chime-sherato...,Sheraton Grand Bengaluru Whitefield Hotel & Co...,Chime - Sheraton Grand Bengaluru Whitefield Ho...,No,Yes,4.3 /5,236,080 49652769,"ITPL Main Road, Whitefield",Bar,"Cocktails, Pizza, Buttermilk",Finger Food,'2500.0',"[('Rated 4.0', 'RATED\n Nice and friendly pla...",[],Pubs and bars,Whitefield


- Use `.head()` to preview the top 5 rows of the dataset.
- Use `.head(n)` to preview the top n rows of the dataset.
- Use `.tail(n)` to preview the bottom n rows of the dataset.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23193 entries, 0 to 23192
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   url                          23193 non-null  object
 1   address                      23193 non-null  object
 2   name                         23193 non-null  object
 3   online_order                 23193 non-null  object
 4   book_table                   23193 non-null  object
 5   rate                         23193 non-null  object
 6   votes                        23193 non-null  int64 
 7   phone                        23193 non-null  object
 8   location                     23193 non-null  object
 9   rest_type                    23193 non-null  object
 10  dish_liked                   23193 non-null  object
 11  cuisines                     23193 non-null  object
 12  approx_cost(for two people)  23193 non-null  object
 13  reviews_list                 23

`df.info()` gives a list of columns with:

- **Name** of columns
- **How many non-null values (blank cells)** each column has.
- **Type of values** in each column - int, float, etc.

**By default**, it shows **Dtype** as `object` for anything other than **int or float**.



---

## **2. Pandas Core Components: DataFrame & Series🧱**

### **1. DataFrame**
A <font color="skyblue">DataFrame</font> is a two-dimensional table with labeled rows and columns. It is the core structure in Pandas, ideal for tabular data like the Zomato dataset.

💡 **Key Features**:
- Can hold data of different types (strings, integers, floats, etc.).
- Labeled rows (**index**) and columns.
- Provides tools for filtering, aggregating, and transforming data.

<img src="https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/065/849/original/download.png?1708489323">


In [7]:
# Viewing a specific column
df['rate']   # Shows the ratings column very similar to Dictionary

0         4.1/5
1         4.1/5
2         3.8/5
3         3.7/5
4         3.8/5
          ...  
23188    3.8 /5
23189    3.9 /5
23190    2.8 /5
23191    2.5 /5
23192    4.3 /5
Name: rate, Length: 23193, dtype: object

In [8]:
# Creating a DataFrame from a dictionary
data = {
    "Restaurant": ["R1", "R2", "R3"],
    "Votes": [120, 240, 150],
    "Rating": [4.1, 4.5, 3.9]
}

restaurant_df = pd.DataFrame(data)
restaurant_df

,Restaurant,Votes,Rating
0,R1,120,4.1
1,R2,240,4.5
2,R3,150,3.9


In [ ]:
restaurant_df.shape

(3, 3)

Similar to Numpy, it gives the **no. of rows and columns**.

In [9]:
type(data['Votes'])

list

In [10]:
type(restaurant_df['Votes'])

pandas.core.series.Series



### **2. Series**
A <font color="green">Series</font> is a one-dimensional labeled array that can hold any data type. Think of it as a **single column of data** in a table.

💡 **Key Features**:
- Data is labeled using an **index**.
- Supports operations like indexing, slicing, and filtering.

<img src="https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/065/850/original/download_%281%29.png?1708489343">  

In [ ]:
# Creating a Series from a list
restaurant_votes = pd.Series([120, 240, 150, 80], index=['R1', 'R2', 'R3', 'R4'])
print(restaurant_votes)

R1    120
R2    240
R3    150
R4     80
dtype: int64


In [ ]:
# Accessing elements in a Series
restaurant_votes['R2']  # Access by index

240

In [ ]:
restaurant_votes.shape

(4,)



---

## **3.🔍 Basic Operations on Columns**

Columns in a Pandas DataFrame represent individual data features.





### **Accessing Columns**
You can access columns using:
- **Dot notation**: `df.column_name`  
- **Bracket notation**: `df["column_name"]`

In [11]:
# Accessing columns
votes = df["votes"]  # Bracket notation
print(votes)

0        775
1        787
2        918
3         88
4        166
        ... 
23188    128
23189     77
23190    161
23191     81
23192    236
Name: votes, Length: 23193, dtype: int64


In [ ]:
rating = df.rate  # Dot notation
print(rating)

0         4.1/5
1         4.1/5
2         3.8/5
3         3.7/5
4         3.8/5
          ...  
23188    3.8 /5
23189    3.9 /5
23190    2.8 /5
23191    2.5 /5
23192    4.3 /5
Name: rate, Length: 23193, dtype: object


In [12]:
df[['url', 'name' ]]  #accessing multiple columns

,url,name
0,https://www.zomato.com/bangalore/jalsa-banasha...,Jalsa
1,https://www.zomato.com/bangalore/spice-elephan...,Spice Elephant
2,https://www.zomato.com/SanchurroBangalore?cont...,San Churro Cafe
3,https://www.zomato.com/bangalore/addhuri-udupi...,Addhuri Udupi Bhojana
4,https://www.zomato.com/bangalore/grand-village...,Grand Village
...,...,...
23188,https://www.zomato.com/bangalore/izakaya-gastr...,Izakaya Gastro Pub
23189,https://www.zomato.com/bangalore/m-bar-bengalu...,M Bar - Bengaluru Marriott Hotel Whitefield
23190,https://www.zomato.com/bangalore/keys-cafe-key...,Keys Cafe - Keys Hotel
23191,https://www.zomato.com/bangalore/bhagini-2-whi...,Bhagini


### **Renaming Columns**
Use the `df.rename()` function to rename columns.


In [13]:
# Renaming columns
df = df.rename(columns={"votes": "Total_Votes", "rate": "Ratings"})

# Displaying renamed DataFrame
df.head()

,url,address,name,online_order,book_table,Ratings,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city)
0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Jalsa,Yes,Yes,4.1/5,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,4.1/5,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari
2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,3.8/5,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari
3,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",Addhuri Udupi Bhojana,No,No,3.7/5,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari
4,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",Grand Village,No,No,3.8/5,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari


💡 **Key Point**: Rename ensures clarity, especially when column names are too short or unclear or have some formatting errors.



### **Adding a New Column**
Add a new column by assigning data to a new column name.

In [14]:
# Adding a new column
df["only_rating"] = [i[:-2] for i in df["Ratings"]]

# Displaying new column
df["only_rating"]

0         4.1
1         4.1
2         3.8
3         3.7
4         3.8
         ... 
23188    3.8 
23189    3.9 
23190    2.8 
23191    2.5 
23192    4.3 
Name: only_rating, Length: 23193, dtype: object

In [15]:
df.head()

,url,address,name,online_order,book_table,Ratings,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating
0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Jalsa,Yes,Yes,4.1/5,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,4.1/5,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1
2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,3.8/5,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari,3.8
3,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",Addhuri Udupi Bhojana,No,No,3.7/5,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari,3.7
4,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",Grand Village,No,No,3.8/5,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari,3.8


In [16]:
# Adding a new column
df['Total_Votes_by_1000']  = df['Total_Votes'] / 1000

# Displaying new column
df['Total_Votes_by_1000']

0        0.775
1        0.787
2        0.918
3        0.088
4        0.166
         ...  
23188    0.128
23189    0.077
23190    0.161
23191    0.081
23192    0.236
Name: Total_Votes_by_1000, Length: 23193, dtype: float64

In [17]:
df["is_good"] = df["Total_Votes"] > 500
df["is_good"]

0         True
1         True
2         True
3        False
4        False
         ...  
23188    False
23189    False
23190    False
23191    False
23192    False
Name: is_good, Length: 23193, dtype: bool

### **Deleting a Column**

Use `df.drop()` to delete one or more columns:
- Set `axis=1` to indicate column deletion.
- Use `inplace=True` to modify the original DataFrame.

In [18]:
# Deleting a column
df.drop("Ratings", axis=1, inplace=True)  # now we do not need the old ratings column

# Displaying updated DataFrame
df.head()

,url,address,name,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Jalsa,Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1,0.775,True
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1,0.787,True
2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari,3.8,0.918,True
3,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",Addhuri Udupi Bhojana,No,No,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari,3.7,0.088,False
4,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",Grand Village,No,No,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari,3.8,0.166,False


### **Unique Values in a Column**

The `df["column_name"].unique()` method lists unique values in a column.

In [19]:
# Finding unique values in the 'Restaurant Rating' column
unique_ratings = df["cuisines"].unique()
print(unique_ratings)

['North Indian, Mughlai, Chinese' 'Chinese, North Indian, Thai'
 'Cafe, Mexican, Italian' ... 'Andhra, Hyderabadi, Biryani'
 'Andhra, North Indian, South Indian' 'Thai, Chinese, Momos']


In [ ]:
unique_ratings = df["only_rating"].unique()
print(unique_ratings)

['4.1' '3.8' '3.7' '4.6' '4.0' '4.2' '3.9' '3.0' '3.6' '2.8' '4.4' '3.1'
 '4.3' '2.6' '3.3' '3.5' '3.8 ' '3.2' '4.5' '2.5' '2.9' '3.4' '2.7' '4.7'
 'N' '2.4' '2.2' '2.3' '4.8' '3.9 ' '4.2 ' '4.0 ' '4.1 ' '2.9 ' '2.7 '
 '2.5 ' '2.6 ' '4.5 ' '4.3 ' '3.7 ' '4.4 ' '4.9' '2.1' '2.0' '1.8' '3.4 '
 '3.6 ' '3.3 ' '4.6 ' '4.9 ' '3.2 ' '3.0 ' '2.8 ' '3.5 ' '3.1 ' '4.8 '
 '2.3 ' '4.7 ' '2.4 ' '2.1 ' '2.2 ' '2.0 ' '1.8 ']


Notice there is a "N" in the rating, might be some typing mistake or might represent null values

The `df["column_name"].nunique()` method lists number of unique values in a column.

In [20]:
number_unique_ratings = df["cuisines"].nunique()
print(number_unique_ratings)

1687


Question: How my different restaurents are present in this dataset?

In [ ]:
num_of_res = df["name"].nunique()
print(num_of_res)

3159


### **Value Counts in a Column**

The `df["column_name"].value_counts()` method counts occurrences of each unique value.

In [ ]:
# Counting occurrences of each unique value
rating_counts = df["only_rating"].value_counts()
rating_counts

,count
only_rating,
3.9,1689
3.9,1549
4.0,1469
4.1,1398
4.0,1391
...,...
2.2,7
2.0,7
2.0,4


In [ ]:
rating_counts["N"]

147

There are total 147 null values in this

Question: Which is the most ordered restaurant in the dataset?

In [ ]:
restaurant_order_counts = df["name"].value_counts()
restaurant_order_counts

,count
name,
Onesta,85
Empire Restaurant,69
KFC,59
McDonald's,57
Pizza Hut,56
...,...
Agraharam Meal,1
Fresh 'N' Grills,1
Coorg Diaries,1


Which city has the most restaurants listed in the dataset?

In [ ]:
city_counts = df['location'].value_counts()
city_counts

,count
location,
Koramangala 5th Block,1783
BTM,1456
Indiranagar,1350
HSR,1162
Jayanagar,1037
...,...
Yelahanka,2
Kanakapura Road,2
West Bangalore,2


Which type of listing (e.g., delivery, dine-out, etc.) is the most common?


In [ ]:
listing_counts = df['listed_in(type)'].value_counts()
listing_counts

,count
listed_in(type),
Delivery,10625
Dine-out,8009
Desserts,1333
Cafes,1047
Drinks & nightlife,890
Buffet,762
Pubs and bars,527




---

## **4. 🧱 Operations on Rows**



### **Understanding Index in a DataFrame**

Every row in a DataFrame has an **index**, which uniquely identifies the row.

Indexing is critical for:
1. Accessing specific rows.
2. Slicing multiple rows.
3. Efficient filtering and data manipulation.

💡 **Key Types of Indexing:**
1. **Implicit Index**: The default integer-based row indices (`0, 1, 2, ...`).
2. **Explicit Index**: Custom labels you assign to rows (e.g., names or IDs).

In [21]:
# Example: Viewing the DataFrame's index
print(df.index)  # Displays the current index of the DataFrame

RangeIndex(start=0, stop=23193, step=1)




### **Implicit vs. Explicit Indexing**

- **Implicit Index**: Automatically generated when creating a DataFrame.
  - Example: `0, 1, 2, 3, ...`
- **Explicit Index**: Custom-defined indices based on the dataset's context.
  - Example: Using `Restaurant` names or `City` names as indices.

You can assign an explicit index using the `set_index()` method:

In [22]:
# Assigning an explicit index
df_explicit = df.set_index("name")
df_explicit

,url,address,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
name,,,,,,,,,,,,,,,,,,
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1,0.775,True
Spice Elephant,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Yes,No,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1,0.787,True
San Churro Cafe,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",Yes,No,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari,3.8,0.918,True
Addhuri Udupi Bhojana,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",No,No,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari,3.7,0.088,False
Grand Village,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",No,No,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari,3.8,0.166,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
Izakaya Gastro Pub,https://www.zomato.com/bangalore/izakaya-gastr...,"2nd Floor, Iona, Virginia Mall, Whitefield Mai...",Yes,Yes,128,+91 7625087121\n+91 8050587483,Whitefield,"Bar, Casual Dining","Beer, Chicken Guntur, Paneer Tikka, Fish, Nood...","North Indian, Continental, Mediterranean",'1200.0',"[('Rated 3.0', ""RATED\n Nice place to hangout...",[],Pubs and bars,Whitefield,3.8,0.128,False
M Bar - Bengaluru Marriott Hotel Whitefield,https://www.zomato.com/bangalore/m-bar-bengalu...,"Bengaluru Marriott Hotel, 75, 8th Road, EPIP A...",No,No,77,080 49435000,Whitefield,"Fine Dining, Bar",Rooftop Ambience,Finger Food,'2000.0',"[('Rated 4.0', 'RATED\n Went there post dinne...",[],Pubs and bars,Whitefield,3.9,0.077,False
Keys Cafe - Keys Hotel,https://www.zomato.com/bangalore/keys-cafe-key...,"Keys Hotel, 6, 1st Phase Industrial Area, Near...",No,No,161,080 39451000\n+91 8884038484,Whitefield,"Casual Dining, Bar","Salads, Coffee, Breakfast Buffet, Halwa, Chick...","Chinese, Continental, North Indian",'1200.0',"[('Rated 3.0', 'RATED\n Place is good not tha...",[],Pubs and bars,Whitefield,2.8,0.161,False


In [23]:
df_explicit.index

Index(['Jalsa', 'Spice Elephant', 'San Churro Cafe', 'Addhuri Udupi Bhojana',
       'Grand Village', 'Timepass Dinner', 'Onesta', 'Penthouse Cafe',
       'Smacznego', 'Café Down The Alley',
       ...
       'BAR BAR', 'The Beer Cafe', 'Pablo's Gastrobar', 'Oliver's Pub & Diner',
       'Smaaash', 'Izakaya Gastro Pub',
       'M Bar - Bengaluru Marriott Hotel Whitefield', 'Keys Cafe - Keys Hotel',
       'Bhagini', 'Chime - Sheraton Grand Bengaluru Whitefield Hotel &...'],
      dtype='object', name='name', length=23193)


### **Slicing Rows**

Rows can be sliced using implicit or explicit indices. Pandas provides two key tools for slicing:

1. **`.iloc`**: Slicing based on implicit integer indices.
2. **`.loc`**: Slicing based on explicit indices.



**Slicing Rows with `.iloc`**

Use `.iloc` for row selection by integer position:

In [24]:
# Slicing rows using `.iloc`
df.iloc[0]   # Access the first row

url                            https://www.zomato.com/bangalore/jalsa-banasha...
address                        942, 21st Main Road, 2nd Stage, Banashankari, ...
name                                                                       Jalsa
online_order                                                                 Yes
book_table                                                                   Yes
Total_Votes                                                                  775
phone                                             080 42297555\r\n+91 9743772233
location                                                            Banashankari
rest_type                                                          Casual Dining
dish_liked                     Pasta, Lunch Buffet, Masala Papad, Paneer Laja...
cuisines                                          North Indian, Mughlai, Chinese
approx_cost(for two people)                                              '800.0'
reviews_list                

In [25]:
df.iloc[1:3] # Access rows from index 1 to 2 (exclusive of 3)

,url,address,name,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1,0.787,True
2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari,3.8,0.918,True


In [26]:
df.iloc[1:10:5] # access multiple non-consecutive rows at same time

,url,address,name,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1,0.787,True
6,https://www.zomato.com/bangalore/onesta-banash...,"2469, 3rd Floor, 24th Cross, Opposite BDA Comp...",Onesta,Yes,Yes,2556,080 48653961\r\n080 48655715,Banashankari,"Casual Dining, Cafe","Farmhouse Pizza, Chocolate Banana, Virgin Moji...","Pizza, Cafe, Italian",'600.0',"[('Rated 5.0', 'RATED\n I personally really l...",[],Cafes,Banashankari,4.6,2.556,True


In [27]:
df.iloc[-1] #last row

url                            https://www.zomato.com/bangalore/chime-sherato...
address                        Sheraton Grand Bengaluru Whitefield Hotel & Co...
name                           Chime - Sheraton Grand Bengaluru Whitefield Ho...
online_order                                                                  No
book_table                                                                   Yes
Total_Votes                                                                  236
phone                                                               080 49652769
location                                              ITPL Main Road, Whitefield
rest_type                                                                    Bar
dish_liked                                          Cocktails, Pizza, Buttermilk
cuisines                                                             Finger Food
approx_cost(for two people)                                             '2500.0'
reviews_list                


### **Slicing Rows with `.loc`**

Use `.loc` for row selection by explicit labels or indices:

In [28]:
# Slicing rows using `.loc`
df_explicit = df.set_index("name")

# Access a single row by explicit label
df_explicit.loc["Jalsa"]

,url,address,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
name,,,,,,,,,,,,,,,,,,
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Delivery,Banashankari,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\r\r\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Dine-out,Banashankari,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Basavanagudi,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Delivery,Basavanagudi,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,804,080 42297555\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Paneer Lajawab, Masala Pa...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n Super ambience\nGreat...",[],Dine-out,Basavanagudi,4.1,0.804,True
Jalsa,https://www.zomato.com/bangalore/jalsa-maratha...,"25/7, Doddannakundi, Outer Ring Road, Mahadeva...",Yes,Yes,2173,+91 7406617555\r\n+91 7406627555,Marathahalli,Casual Dining,"Mutton Curry, Stuffed Mushroom, Laccha Paratha...","North Indian, Mughlai",'1500.0',"[('Rated 4.0', 'RATED\n A pretty decent place...",[],Delivery,Brookefield,4.2,2.173,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,783,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Jayanagar,4.1,0.783,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,783,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Delivery,Jayanagar,4.1,0.783,True


In [29]:
# Access multiple rows by explicit label
df_explicit.loc[["Jalsa", "Tuk-Tuk"]]

,url,address,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
name,,,,,,,,,,,,,,,,,,
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Delivery,Banashankari,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\r\r\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Dine-out,Banashankari,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Basavanagudi,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Delivery,Basavanagudi,4.1,0.775,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,804,080 42297555\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Paneer Lajawab, Masala Pa...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n Super ambience\nGreat...",[],Dine-out,Basavanagudi,4.1,0.804,True
Jalsa,https://www.zomato.com/bangalore/jalsa-maratha...,"25/7, Doddannakundi, Outer Ring Road, Mahadeva...",Yes,Yes,2173,+91 7406617555\r\n+91 7406627555,Marathahalli,Casual Dining,"Mutton Curry, Stuffed Mushroom, Laccha Paratha...","North Indian, Mughlai",'1500.0',"[('Rated 4.0', 'RATED\n A pretty decent place...",[],Delivery,Brookefield,4.2,2.173,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,783,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Jayanagar,4.1,0.783,True
Jalsa,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Yes,Yes,783,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Delivery,Jayanagar,4.1,0.783,True


**Key Methods for Row Operations**

1. **View Current Index**: `df.index`
2. **Change Index**: `df.set_index("column_name")`
3. **Access Rows by Position**: `df.iloc[]`
4. **Access Rows by Label**: `df.loc[]`
5. **Slice Rows**:
   - Implicit Index: `df.iloc[start:end]`
   - Explicit Index: `df.loc["label_start":"label_end"]`



**How can we reset our indices back to integers?**



In [30]:
df.reset_index()

,index,url,address,name,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
0,0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Jalsa,Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1,0.775,True
1,1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1,0.787,True
2,2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari,3.8,0.918,True
3,3,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",Addhuri Udupi Bhojana,No,No,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari,3.7,0.088,False
4,4,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",Grand Village,No,No,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari,3.8,0.166,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
23188,23188,https://www.zomato.com/bangalore/izakaya-gastr...,"2nd Floor, Iona, Virginia Mall, Whitefield Mai...",Izakaya Gastro Pub,Yes,Yes,128,+91 7625087121\n+91 8050587483,Whitefield,"Bar, Casual Dining","Beer, Chicken Guntur, Paneer Tikka, Fish, Nood...","North Indian, Continental, Mediterranean",'1200.0',"[('Rated 3.0', ""RATED\n Nice place to hangout...",[],Pubs and bars,Whitefield,3.8,0.128,False
23189,23189,https://www.zomato.com/bangalore/m-bar-bengalu...,"Bengaluru Marriott Hotel, 75, 8th Road, EPIP A...",M Bar - Bengaluru Marriott Hotel Whitefield,No,No,77,080 49435000,Whitefield,"Fine Dining, Bar",Rooftop Ambience,Finger Food,'2000.0',"[('Rated 4.0', 'RATED\n Went there post dinne...",[],Pubs and bars,Whitefield,3.9,0.077,False
23190,23190,https://www.zomato.com/bangalore/keys-cafe-key...,"Keys Hotel, 6, 1st Phase Industrial Area, Near...",Keys Cafe - Keys Hotel,No,No,161,080 39451000\n+91 8884038484,Whitefield,"Casual Dining, Bar","Salads, Coffee, Breakfast Buffet, Halwa, Chick...","Chinese, Continental, North Indian",'1200.0',"[('Rated 3.0', 'RATED\n Place is good not tha...",[],Pubs and bars,Whitefield,2.8,0.161,False
23191,23191,https://www.zomato.com/bangalore/bhagini-2-whi...,"139/C1, Next To GR Tech Park, Pattandur Agraha...",Bhagini,No,No,81,080 65951222,Whitefield,"Casual Dining, Bar","Biryani, Andhra Meal","Andhra, South Indian, Chinese, North Indian",'800.0',"[('Rated 4.0', 'RATED\n A fine place to chill...",[],Pubs and bars,Whitefield,2.5,0.081,False


Notice that it's creating a new column `index`.

**How can we reset our index without creating this new column?**

In [31]:
df.reset_index(drop=True) # by using drop=True we can prevent creating a new column

,url,address,name,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Jalsa,Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1,0.775,True
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1,0.787,True
2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari,3.8,0.918,True
3,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",Addhuri Udupi Bhojana,No,No,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari,3.7,0.088,False
4,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",Grand Village,No,No,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari,3.8,0.166,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
23188,https://www.zomato.com/bangalore/izakaya-gastr...,"2nd Floor, Iona, Virginia Mall, Whitefield Mai...",Izakaya Gastro Pub,Yes,Yes,128,+91 7625087121\n+91 8050587483,Whitefield,"Bar, Casual Dining","Beer, Chicken Guntur, Paneer Tikka, Fish, Nood...","North Indian, Continental, Mediterranean",'1200.0',"[('Rated 3.0', ""RATED\n Nice place to hangout...",[],Pubs and bars,Whitefield,3.8,0.128,False
23189,https://www.zomato.com/bangalore/m-bar-bengalu...,"Bengaluru Marriott Hotel, 75, 8th Road, EPIP A...",M Bar - Bengaluru Marriott Hotel Whitefield,No,No,77,080 49435000,Whitefield,"Fine Dining, Bar",Rooftop Ambience,Finger Food,'2000.0',"[('Rated 4.0', 'RATED\n Went there post dinne...",[],Pubs and bars,Whitefield,3.9,0.077,False
23190,https://www.zomato.com/bangalore/keys-cafe-key...,"Keys Hotel, 6, 1st Phase Industrial Area, Near...",Keys Cafe - Keys Hotel,No,No,161,080 39451000\n+91 8884038484,Whitefield,"Casual Dining, Bar","Salads, Coffee, Breakfast Buffet, Halwa, Chick...","Chinese, Continental, North Indian",'1200.0',"[('Rated 3.0', 'RATED\n Place is good not tha...",[],Pubs and bars,Whitefield,2.8,0.161,False
23191,https://www.zomato.com/bangalore/bhagini-2-whi...,"139/C1, Next To GR Tech Park, Pattandur Agraha...",Bhagini,No,No,81,080 65951222,Whitefield,"Casual Dining, Bar","Biryani, Andhra Meal","Andhra, South Indian, Chinese, North Indian",'800.0',"[('Rated 4.0', 'RATED\n A fine place to chill...",[],Pubs and bars,Whitefield,2.5,0.081,False


Great!

Now let's do this in place.

In [32]:
df.reset_index(drop=True, inplace=True)

In [33]:
df.head()

,url,address,name,online_order,book_table,Total_Votes,phone,location,rest_type,dish_liked,cuisines,approx_cost(for two people),reviews_list,menu_item,listed_in(type),listed_in(city),only_rating,Total_Votes_by_1000,is_good
0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ...",Jalsa,Yes,Yes,775,080 42297555\r\n+91 9743772233,Banashankari,Casual Dining,"Pasta, Lunch Buffet, Masala Papad, Paneer Laja...","North Indian, Mughlai, Chinese",'800.0',"[('Rated 4.0', 'RATED\n A beautiful place to ...",[],Buffet,Banashankari,4.1,0.775,True
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...",Spice Elephant,Yes,No,787,080 41714161,Banashankari,Casual Dining,"Momos, Lunch Buffet, Chocolate Nirvana, Thai G...","Chinese, North Indian, Thai",'800.0',"[('Rated 4.0', 'RATED\n Had been here for din...",[],Buffet,Banashankari,4.1,0.787,True
2,https://www.zomato.com/SanchurroBangalore?cont...,"1112, Next to KIMS Medical College, 17th Cross...",San Churro Cafe,Yes,No,918,+91 9663487993,Banashankari,"Cafe, Casual Dining","Churros, Cannelloni, Minestrone Soup, Hot Choc...","Cafe, Mexican, Italian",'800.0',"[('Rated 3.0', ""RATED\n Ambience is not that ...",[],Buffet,Banashankari,3.8,0.918,True
3,https://www.zomato.com/bangalore/addhuri-udupi...,"1st Floor, Annakuteera, 3rd Stage, Banashankar...",Addhuri Udupi Bhojana,No,No,88,+91 9620009302,Banashankari,Quick Bites,Masala Dosa,"South Indian, North Indian",'300.0',"[('Rated 4.0', ""RATED\n Great food and proper...",[],Buffet,Banashankari,3.7,0.088,False
4,https://www.zomato.com/bangalore/grand-village...,"10, 3rd Floor, Lakshmi Associates, Gandhi Baza...",Grand Village,No,No,166,+91 8026612447\r\n+91 9901210005,Basavanagudi,Casual Dining,"Panipuri, Gol Gappe","North Indian, Rajasthani",'600.0',"[('Rated 4.0', 'RATED\n Very good restaurant ...",[],Buffet,Banashankari,3.8,0.166,False




---

## **5. 🎯 Working with Both Rows and Columns**

When analyzing data, you often need to work with both rows and columns simultaneously. Pandas provides powerful tools for selecting, slicing, and filtering data across multiple axes.

💡 **Key Tools for Selection**:
1. **`.iloc`**: Select rows and columns by integer position.
2. **`.loc`**: Select rows and columns by labels.


### **Using `.iloc` for Rows and Columns**

The `.iloc` method allows you to:
1. Select specific rows and columns by their integer indices.
2. Slice rows and columns together.

#### Example: Selecting Specific Rows and Columns

In [ ]:
# Select the first two rows and the first two columns
df.iloc[0:2, 0:2]

,url,address
0,https://www.zomato.com/bangalore/jalsa-banasha...,"942, 21st Main Road, 2nd Stage, Banashankari, ..."
1,https://www.zomato.com/bangalore/spice-elephan...,"2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ..."


In [ ]:
# Select the last two rows and the last column
df.iloc[-2:, -1]

,is_good
23191,False
23192,False


**2. Using `.iloc` to Modify Data**

In [ ]:
# Example DataFrame
data = {
    "Restaurant": ["R1", "R2", "R3", "R4"],
    "Votes": [120, 240, 150, 80],
    "Rating": [4.1, 4.5, 3.9, 4.0]
}

sample_df = pd.DataFrame(data)

# Assigning new values to multiple rows and columns
sample_df.iloc[1:3, 1:3] = [[300, 4.7], [180, 4.2]]

print("DataFrame after modifying rows and columns:")
print(sample_df)

DataFrame after modifying rows and columns:
  Restaurant  Votes  Rating
0         R1    120     4.1
1         R2    300     4.7
2         R3    180     4.2
3         R4     80     4.0


**3. Using `.loc` to Select Rows and Columns**

The `.loc` method allows you to:
- Filter rows based on conditions.
- Update specific rows and columns by labels.

In [ ]:
# Setting 'Restaurant' as the index
df_explicit = sample_df.set_index("Restaurant")

# Select specific rows ('R2' and 'R3') and specific columns ('Votes')
print(df_explicit.loc[["R2", "R3"], ["Votes"]])
print("==="*10)

# Updating the 'Votes' for restaurant 'R2'
df_explicit.loc["R2", "Votes"] = 350
print("Updated DataFrame:")
df_explicit

            Votes
Restaurant       
R2            300
R3            180
Updated DataFrame:


,Votes,Rating
Restaurant,,
R1,120,4.1
R2,350,4.7
R3,180,4.2
R4,80,4.0


**4. Using `.loc` for Conditional Updates**

In [ ]:
# Update 'Rating' for all restaurants with 'Votes' greater than 200
df_explicit.loc[df_explicit["Votes"] > 200, "Rating"] = 4.8
print("DataFrame after conditional update:")
df_explicit

DataFrame after conditional update:


,Votes,Rating
Restaurant,,
R1,120,4.1
R2,350,4.8
R3,180,4.2
R4,80,4.0


### ✂️ Combining DataFrames with `concat`

The `concat` function allows you to combine multiple DataFrames:
1. **Row-wise Concatenation**: Add rows from another DataFrame.
2. **Column-wise Concatenation**: Add new columns from another DataFrame.



**1. Row-wise Concatenation**

You can stack two DataFrames vertically to combine their rows:


In [ ]:
# Creating a second DataFrame
data2 = {
    "Restaurant": ["R5", "R6"],
    "Votes": [90, 110],
    "Rating": [3.8, 4.3]
}

df2 = pd.DataFrame(data2)

# Concatenating row-wise
combined_df = pd.concat([sample_df, df2], ignore_index=True)
print("Combined DataFrame (Row-wise):")
print(combined_df)

Combined DataFrame (Row-wise):
  Restaurant  Votes  Rating
0         R1    120     4.1
1         R2    300     4.7
2         R3    180     4.2
3         R4     80     4.0
4         R5     90     3.8
5         R6    110     4.3


**2. Column-wise Concatenation**

You can stack DataFrames side-by-side to add new columns:

In [ ]:
# Creating a DataFrame with new columns
new_columns = {
    "Delivery Time (mins)": [30, 25, 35, 20, 40, 28]
}

df_new_cols = pd.DataFrame(new_columns)

# Concatenating column-wise
final_df = pd.concat([combined_df, df_new_cols], axis=1)
print("Final DataFrame (Column-wise):")
final_df

Final DataFrame (Column-wise):


,Restaurant,Votes,Rating,Delivery Time (mins)
0,R1,120,4.1,30
1,R2,300,4.7,25
2,R3,180,4.2,35
3,R4,80,4.0,20
4,R5,90,3.8,40
5,R6,110,4.3,28


### 📝 Practice Questions

1. **Using `.iloc`**, retrieve the first two rows and the last column.


In [ ]:
# 1. Using `.iloc` to retrieve first two rows and last column
sample_df.iloc[:2, -1]

,Rating
0,4.1
1,4.7


2. **Using `.loc`**, update the 'Rating' for restaurants with votes below 150 to 4.0.

In [ ]:
# 2. Using `.loc` to update 'Rating' for restaurants with votes < 150
sample_df.loc[sample_df["Votes"] < 150, "Rating"] = 4.0
sample_df

,Restaurant,Votes,Rating
0,R1,120,4.0
1,R2,300,4.7
2,R3,180,4.2
3,R4,80,4.0


3. Add a new row to the DataFrame for a restaurant 'R7' with 100 votes and a rating of 3.7.

In [ ]:
# 3. Adding a new row
new_row = pd.DataFrame([["R7", 100, 3.7]], columns=sample_df.columns)
sample_df = pd.concat([sample_df, new_row], ignore_index=True)
sample_df

,Restaurant,Votes,Rating
0,R1,120,4.0
1,R2,300,4.7
2,R3,180,4.2
3,R4,80,4.0
4,R7,100,3.7


4. Concatenate two DataFrames:
   - One with restaurant data.
   - One with a column showing average delivery time.

In [ ]:
# 4. Concatenating two DataFrames
df_with_time = pd.DataFrame({"Delivery Time (mins)": [30, 40, 20, 35, 25]})
final_combined_df = pd.concat([sample_df, df_with_time], axis=1)
final_combined_df

,Restaurant,Votes,Rating,Delivery Time (mins)
0,R1,120,4.0,30
1,R2,300,4.7,40
2,R3,180,4.2,20
3,R4,80,4.0,35
4,R7,100,3.7,25


5. Find the updated 'Rating' for restaurants with votes greater than 200 after combining two DataFrames.

In [ ]:
# 5. Finding updated 'Rating' for restaurants with votes > 200
updated_ratings = final_combined_df.loc[final_combined_df["Votes"] > 200, "Rating"]
print(updated_ratings)

1    4.7
Name: Rating, dtype: float64




---

## **6. 🗑️ Handling Duplicate Records in a DataFrame**

In real-world datasets, duplicate records are common and can affect analysis. Pandas provides simple methods to identify and remove duplicates efficiently.


### **Key Methods for Handling Duplicates**

### **1. Identifying Duplicate Rows with `df.duplicated()`**

The `df.duplicated()` method:
- Returns `True` for duplicate rows (excluding the first occurrence by default).
- You can customize its behavior using the `keep` parameter:
  - `keep='first'` (default): Marks all duplicates except the first occurrence.
  - `keep='last'`: Marks all duplicates except the last occurrence.
  - `keep=False`: Marks **all** duplicates as `True`.

In [ ]:
# Example DataFrame
data = {
    "Restaurant": ["R1", "R2", "R3", "R2", "R4", "R1"],
    "Votes": [120, 240, 150, 240, 80, 120],
    "Rating": [4.1, 4.5, 3.9, 4.5, 4.0, 4.1]
}

sample_df = pd.DataFrame(data)
sample_df

,Restaurant,Votes,Rating
0,R1,120,4.1
1,R2,240,4.5
2,R3,150,3.9
3,R2,240,4.5
4,R4,80,4.0
5,R1,120,4.1


In [ ]:
# Identifying duplicate rows
print("Duplicate rows (keep='first'):")
print(sample_df.duplicated())  # Default behavior

Duplicate rows (keep='first'):
0    False
1    False
2    False
3     True
4    False
5     True
dtype: bool


In [ ]:
print("Duplicate rows (keep='last'):")
print(sample_df.duplicated(keep="last"))

Duplicate rows (keep='last'):
0     True
1     True
2    False
3    False
4    False
5    False
dtype: bool


In [ ]:
print("Duplicate rows (keep=False):")
print(sample_df.duplicated(keep=False))

Duplicate rows (keep=False):
0     True
1     True
2    False
3     True
4    False
5     True
dtype: bool


**Key Points**
- **Default (`keep='first'`)**: Treats the first occurrence of a duplicate as unique.
- **`keep='last'`**: Treats the last occurrence of a duplicate as unique.
- **`keep=False`**: Marks all duplicates as `True`.

### **2. Removing Duplicate Rows**

The `df.drop_duplicates()` method removes duplicate rows based on the specified conditions:
- By default, it keeps the first occurrence of each duplicate row.
- You can use the `keep` parameter to change this behavior.

In [ ]:
# Removing duplicates (keep='first')
df_no_duplicates_first = sample_df.drop_duplicates(keep="first")
print("DataFrame after removing duplicates (keep='first'):")
print(df_no_duplicates_first)

DataFrame after removing duplicates (keep='first'):
  Restaurant  Votes  Rating
0         R1    120     4.1
1         R2    240     4.5
2         R3    150     3.9
4         R4     80     4.0


In [ ]:
# Removing duplicates (keep='last')
df_no_duplicates_last = sample_df.drop_duplicates(keep="last")
print("DataFrame after removing duplicates (keep='last'):")
print(df_no_duplicates_last)

DataFrame after removing duplicates (keep='last'):
  Restaurant  Votes  Rating
2         R3    150     3.9
3         R2    240     4.5
4         R4     80     4.0
5         R1    120     4.1


In [ ]:
# Removing all duplicates (keep=False)
df_no_duplicates_all = sample_df.drop_duplicates(keep=False)
print("DataFrame after removing all duplicates (keep=False):")
print(df_no_duplicates_all)

DataFrame after removing all duplicates (keep=False):
  Restaurant  Votes  Rating
2         R3    150     3.9
4         R4     80     4.0


### **3. Removing Duplicates Based on Specific Columns**

Sometimes, you might want to check duplicates only for certain columns. Use the `subset` parameter in `df.drop_duplicates()` to specify the columns to consider for duplication.

- Use `subset` in `df.drop_duplicates()` to check duplicates for specific columns.

In [ ]:
# Removing duplicates based on the 'Votes' column
df_no_duplicates_subset = sample_df.drop_duplicates(subset=["Votes"], keep="first")
print("DataFrame after removing duplicates based on 'Votes':")
print(df_no_duplicates_subset)

DataFrame after removing duplicates based on 'Votes':
  Restaurant  Votes  Rating
0         R1    120     4.1
1         R2    240     4.5
2         R3    150     3.9
4         R4     80     4.0


In [ ]:
# Removing duplicates based on 'Restaurant' and 'Votes'
df_no_duplicates_multi_subset = sample_df.drop_duplicates(subset=["Restaurant", "Votes"], keep="first")
print("DataFrame after removing duplicates based on 'Restaurant' and 'Votes':")
print(df_no_duplicates_multi_subset)

DataFrame after removing duplicates based on 'Restaurant' and 'Votes':
  Restaurant  Votes  Rating
0         R1    120     4.1
1         R2    240     4.5
2         R3    150     3.9
4         R4     80     4.0




---

## **7. 📊 Pandas Built-In Operations: Aggregate Functions**


In [ ]:
# Sum of votes
print("Total Votes:", df["Total_Votes"].sum())

Total Votes: 13940720


In [ ]:
# Count of ratings
print("Number of Ratings:", df["only_rating"].count())

Number of Ratings: 23193


In [ ]:
# Select relevant columns and preprocess
df['approx_cost(for two people)'] = df['approx_cost(for two people)'].str.replace(',', '')
df['approx_cost(for two people)'] = df['approx_cost(for two people)'].str.replace("'", '')
df['approx_cost(for two people)'] = df['approx_cost(for two people)'].astype(float)

In [ ]:
# Average cost for two
print("Average Cost for Two:", df["approx_cost(for two people)"].mean())

Average Cost for Two: 755.90436769715


### **Applying Aggregate Functions on Multiple Columns**

You can also apply aggregate functions across multiple columns using the `.agg()` method.

#### Example: Aggregating Across Columns

In [ ]:
# Aggregating multiple columns
aggregates = df[["Total_Votes", "approx_cost(for two people)"]].agg(["sum", "mean", "min", "max"])
print("Aggregates for Votes and Cost for Two:")
aggregates

Aggregates for Votes and Cost for Two:


,Total_Votes,approx_cost(for two people)
sum,1.394072e+07,1.753169e+07
mean,6.010745e+02,7.559044e+02
min,0.000000e+00,4.000000e+01
max,1.683200e+04,6.000000e+03



---

## **8. 🔀 Sorting Values in Pandas**

### **Sorting by a Single Column**



In [ ]:
# Sorting rows by the "Votes" column in ascending order
sorted_votes_asc = sample_df.sort_values(by="Votes", ascending=True)
print("DataFrame sorted by Votes (ascending):")
print(sorted_votes_asc)

DataFrame sorted by Votes (ascending):
  Restaurant  Votes  Rating
4         R4     80     4.0
0         R1    120     4.1
5         R1    120     4.1
2         R3    150     3.9
1         R2    240     4.5
3         R2    240     4.5


In [ ]:
# Sorting rows by the "Votes" column in descending order
sorted_votes_desc = sample_df.sort_values(by="Votes", ascending=False)
print("DataFrame sorted by Votes (descending):")
print(sorted_votes_desc)

DataFrame sorted by Votes (descending):
  Restaurant  Votes  Rating
1         R2    240     4.5
3         R2    240     4.5
2         R3    150     3.9
0         R1    120     4.1
5         R1    120     4.1
4         R4     80     4.0




### **Sorting by Multiple Columns**

**Example: Sorting by "Votes" and "Rating"**

You can sort rows based on multiple columns by passing a list of column names.


In [ ]:
# Sorting by Votes (descending) and then by Rating (ascending)
sorted_multi = sample_df.sort_values(by=["Votes", "Rating"], ascending=[False, True])
print("DataFrame sorted by Votes (descending) and Rating (ascending):")
sorted_multi

DataFrame sorted by Votes (descending) and Rating (ascending):


,Restaurant,Votes,Rating
1,R2,240,4.5
3,R2,240,4.5
2,R3,150,3.9
0,R1,120,4.1
5,R1,120,4.1
4,R4,80,4.0


**Some Practice Questions**

1. Sort the DataFrame by "Rating" in descending order.
2. Find the restaurant with the highest rating after sorting.
3. Sort the DataFrame first by "Cost for Two" (ascending) and then by "Votes" (descending).
4. Create a new sorted DataFrame based on "Votes" and "Rating" and display the top 3 rows.



---

